pkg/pipeline/source/sdk.go

```diff
@@ -145,78 +145,94 @@ func (s *SDKSource) joinRoom() error {
 		OnDisconnected: s.onDisconnected,
 	}
 
-	s.room = lksdk.CreateRoom(cb)
 	logger.Debugw("connecting to room")
+	s.room = lksdk.CreateRoom(cb)
 	if err := s.room.JoinWithToken(s.WsUrl, s.Token, lksdk.WithAutoSubscribe(false)); err != nil {
 		return err
 	}
 
 	var fileIdentifier string
-	tracks := make(map[string]struct{})
-
+	var err error
 	switch s.RequestType {
 	case types.RequestTypeTrackComposite:
 		fileIdentifier = s.Info.RoomName
+
+		tracks := make(map[string]struct{})
 		if s.AudioEnabled {
 			tracks[s.AudioTrackID] = struct{}{}
 		}
 		if s.VideoEnabled {
 			tracks[s.VideoTrackID] = struct{}{}
 		}
+		err = s.awaitTracks(tracks)
 
 	case types.RequestTypeTrack:
 		fileIdentifier = s.TrackID
-		tracks[s.TrackID] = struct{}{}
-	}
 
-	numTracks := len(tracks)
-	s.errors = make(chan error, numTracks)
-	if err := s.subscribeToTracks(tracks); err != nil {
+		err = s.awaitTracks(map[string]struct{}{s.TrackID: {}})
+	}
+	if err != nil {
 		return err
 	}
 
-	for i := 0; i < numTracks; i++ {
-		if err := <-s.errors; err != nil {
-			return err
-		}
+	if err = s.UpdateInfoFromSDK(fileIdentifier, s.filenameReplacements); err != nil {
+		logger.Errorw("could not update file params", err)
+		return err
 	}
 
-	s.initialized.Break()
+	return nil
+}
 
-	if err := s.UpdateInfoFromSDK(fileIdentifier, s.filenameReplacements); err != nil {
-		logger.Errorw("could not update file params", err)
+func (s *SDKSource) awaitTracks(expecting map[string]struct{}) error {
+	trackCount := len(expecting)
+	s.errors = make(chan error, trackCount)
+
+	deadline := time.After(subscriptionTimeout)
```

Comment:
nit: would a context be a more self contained alternative here?

Reply:
will move that over with participant composite